app/models/interactive_run_state.rb

```diff
@@ -145,8 +177,17 @@ def to_json(arg=nil)
     arg ? original_json(arg) : answer_json
   end
 
-  def to_runtime_json()
-    self.to_json({methods: [:linked_state, :has_linked_interactive, :run_remote_endpoint]})
+  def to_runtime_json(host)
+    hash = self.as_json
+    hash[:has_linked_interactive] = has_linked_interactive
+    hash[:linked_state] = linked_state
+    hash[:run_remote_endpoint] = run_remote_endpoint
+    hash[:all_linked_states] = all_linked_states(host)
+    hash[:interactive_state_url] = interactive_state_url(host)
+    hash[:page_index] = page && page.index_in_activity
+    hash[:page_name] = page && page.name
+    hash[:activity_name] = activity && activity.name
+    hash.to_json
```

Comment:
Since this is roughly the same as what is returned for each linked_state_info. It would be better to unify these two somewhat.  One difference is data vs raw_data. Another is that this top level interactive state includes all of the properties while the linked state only selects a few. And finally this one is recursive while the other one isn't. I probably missed some differences. But here is a suggestion based on those differences:

- use `raw_data` for the linked interactive state too 
- it would be safer to explicitly define which properties of the interactive run state are sent down, so limiting this to just created_at, updated_at, raw_data, and interactive_id seems fine to me. However it should be backward compatible, so perhaps the current doc-store interactive expects some of the other properties (run_id, learner_url, or key).
- passing a `include_linked_states` param to this method would solve the recursive issue

Reply:
Done. You should see that in the last commit. It was a bit tricky, as `to_runtime_json` is used both by API and `initInteractive` post message. I've tried to include reasonable params that were enforced either by `initInteractive` format, API specs or common sense.
I've also refactored `initInteractive` message to make use of the more consisten approach.